### ISSUE TITLE
Built-in clipboard manager

### ISSUE THREAD

**Issue by YaLTeR** [YaLTeR/niri#112]
Store text from clients. Basically exactly what Mutter does, it works perfectly there.

**Comment by SalmanFarooqShiekh** [comment]
Isn't this already done?

**Comment by YaLTeR** [comment]
Nope

**Comment by salman-farooq-sh** [comment]
what does this involve then?

**Comment by YaLTeR** [comment]
Storing text when it is copied by clients. So that you can e.g. copy text from Alacritty, close Alacritty, paste it somewhere else. Normally you wouldn't be able to, since clients must remain alive to serve clipboard contents.

**Comment by baduhai** [comment]
The way kwin handles the clipboard manager may also be interesting to have a look at, as it also stores files in the clipboard history.

**Comment by YaLTeR** [comment]
Not sure I want the whole clipboard history in niri, there are 3rd party clipboard managers for that. But storing 1 text selection, yes.

**Comment by MauroGuida** [comment]
Something similar to [wl-clip-persist](https://github.com/Linus789/wl-clip-persist) should do the job

**Comment by PanAeon** [comment]
@YaLTeR I'd like to work on this

**Comment by eveeifyeve** [comment]
Would implementing wl-clipboard-rs inside of niri would work?

**Comment by cies** [comment]
I also miss KWin's clipboard history... If someone has a setup that works, I'd love to see some config for it.

**Comment by HigherOrderLogic** [comment]
> I also miss KWin's clipboard history... If someone has a setup that work, I'd love to see some config for it.

```sh
cliphist list | fuzzel  --match-mode fzf --dmenu | cliphist decode | wl-copy
```

I use a keybind to spawn this. Works for me.

**Comment by akileshas** [comment]
> I also miss KWin's clipboard history... If someone has a setup that works, I'd love to see some config for it.

hey. check out these references.
- https://blog.ricardof.dev/fixing-clipboard-persistence-nixos-wayland/
- https://github.com/sentriz/cliphist/

TL;DR:
install the following packages - `wl-clipboard`, `wl-clip-persist`, `cliphist`.
and just add these startup spawn process in your niri config,
```kdl
spawn-sh-at-startup "wl-clip-persist --clipboard both"
spawn-sh-at-startup "wl-paste --type text --watch cliphist store"
spawn-sh-at-startup "wl-paste --type image --watch cliphist store"
spawn-sh-at-startup "wl-paste --primary --type text --watch cliphist store"
```
then, integrate it with your dmenu of choice. mine is fuzzel.
in fuzzel, you can add a bind in your niri config like,
```kdl
binds {
  Mod+V repeat=false { spawn-sh "cliphist list | fuzzel --with-nth 2 --dmenu | cliphist decode | wl-copy"; }
}
```

**Comment by YaLTeR** [comment]
I haven't written here I guess, but some months ago I started working on the niri test suite to be able to test this, and it turned out to be quite involved with quite a lot of effort, so that stalled until I get back to it at some point.
